### PR TITLE
Initdb refactor

### PR DESCRIPTION
**Pull request by dlowrey**
Refactored the `initdb.py` script to make it more concise and clean. 

Changes include:

- Reorganization of functions to connect, initialize database, and insert data
- Correct implementation of transactions with rollbacks when inserting data
- Optimization inserting data with `executemany()` instead of `execute()`
- Noticeable execution time increse
- Overall cleaned up methods and comments

### CODE REVIEW

File: dbmanager.py [hunk 1]
```diff
@@ -1 +1,65 @@
-#  script to handle MySQL database
+from exceptions import InputError, MySqlError
```
Comment (asclines) [inline dbmanager.py:1]: Pylint output: 
> ************* Module dbmanager
C:  8, 0: Trailing whitespace (trailing-whitespace)
C: 11, 0: Trailing whitespace (trailing-whitespace)
C: 14, 0: Trailing whitespace (trailing-whitespace)
C: 24, 0: Wrong continued indentation (remove 1 space).
                                              password=password,
                                             |^ (bad-continuation)
C: 25, 0: Wrong continued indentation (remove 1 space).
                                              database=DB_NAME)
                                             |^ (bad-continuation)
C: 34, 0: Trailing whitespace (trailing-whitespace)
C: 37, 0: Trailing whitespace (trailing-whitespace)
C: 40, 0: Trailing whitespace (trailing-whitespace)
C: 53, 0: Wrong continued indentation (remove 2 spaces).
                             args=err.args)
                           | ^ (bad-continuation)
~C:  1, 0: Missing module docstring (missing-docstring)~

Ignored crossed out ones.

File: initdb.py [hunk 2]
```diff
@@ -1,5 +1,3 @@
-# Initializes a MySQL Database
-
 import os
```
Comment (asclines) [inline initdb.py:1]: Pylint output:
> ************* Module initdb
C: 85, 0: Trailing whitespace (trailing-whitespace)
C:140, 0: Trailing whitespace (trailing-whitespace)
C:144, 0: Trailing whitespace (trailing-whitespace)
C:150, 0: Trailing whitespace (trailing-whitespace)
~C:  1, 0: Missing module docstring (missing-docstring)~ 

Ignore crossed out ones.

File: dbmanager.py [hunk 3]
```diff
@@ -1 +1,65 @@
...
+def init_connection(username=None, password=None):
+    """Establish a connection to database
+    
+    :arg username: the username required to access the server
+    :arg password: the password required to access the server
+    
+    :returns None if no username or password was provided,
+            a valid connection (if one was found) otherwise
+    
+    :raises An InputError (likely because no username or password was provided)
```
Comment (asclines) [inline dbmanager.py]: Is there anytime this actually happens? Wouldn't an exception just be thrown?

File: dbmanager.py [hunk 4]
```diff
@@ -1 +1,65 @@
...
+    """
+    cursor = None
+    if sql and connection:
+        cursor = connection.cursor()
+        try:
+            return cursor.execute(sql)
+
+        except mysql.connector.Error  as err:
+            raise \
+                MySqlError(message='There was a problem executing your query.',
```
Comment (asclines) [inline dbmanager.py]: Is it possible to include what the SQL statement was in this exception?

Or is this handled by `args`?

File: dbmanager.py [hunk 5]
```diff
@@ -1 +1,65 @@
...
+        connection = mysql.connector.connect(user=username,
+                                              password=password,
+                                              database=DB_NAME)
+    except mysql.connector.Error as err:
+        raise InputError(message='There was an error connecting to the server.',
+                         args=err.args)
+    return connection
+
+
+def exec_sql(connection, sql=None):
```
Comment (asclines) [inline dbmanager.py]: Why `sql=None`? Is there a scenario where we don't want SQL to be supplied?

File: initdb.py [hunk 6]
```diff
@@ -67,107 +65,125 @@
...
+                                                  'income-data.csv')))
+DATASETS['starbucks'] = csv.reader(open(os.path.join('datasets',
+                                                     'starbucks.csv')))
+DATASETS['diversity'] = csv.reader(open(os.path.join('datasets',
+                                                     'diversity.csv')))
+DATASETS['locations'] = csv.reader(open(os.path.join('datasets',
+                                                     'locations.csv')))
+
+
+def init_connections(username=None, password=None):
```
Comment (asclines) [inline initdb.py:80]: Is this the same as the `init_connection` method in `dbmanager.py`?

**Review by asclines — CHANGES_REQUESTED** [review]
I liked the refactoring done on `initdb.py`

File: initdb.py (hunk 6 above)
Comment (dlowrey) [inline initdb.py]: Oh shoot I didn't realize it pushed my dbmanager changes too. I wasn't near finished with dbmanager

**Comment by dlowrey** [comment]
Oh shoot I didn't realize it pushed my dbmanager changes too. I wasn't near finished with dbmanager

**Comment by dlowrey** [comment]
I will clean up `dbmanager.py` to just include the connecting method and the method signature for running SQL queries and we can implement them later.

File: initdb.py (hunk 6 above)
Comment (dlowrey) [reply initdb.py:80]: Yes, I am guessing that we should instead have `initdb.py` get its connection by calling the `init_connectoin` method in `dbmanager` instead, and have every module that is going to connect do this as well.

**Comment by asclines** [comment]
@dlowrey implement what later?

**Comment by dlowrey** [comment]
@asclines I decided that the functionality was already there so I just cleaned it up.

**Review by asclines — APPROVED** [review]

File: initdb.py (hunk 2 above)
Comment (dlowrey) [reply initdb.py:1]: @asclines It really doesn't like the whitespace left between the Doc-string comments?

**Comment by asclines** [comment]
Approving with the expectation that more changes will come later.

**Pull request merged**